File: Kernel Simulator/Misc/Screensaver/Displays/FlashText.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace Misc.Screensaver.Displays
    Public Module FlashTextSettings

        Private _flashText255Colors As Boolean
        Private _flashTextTrueColor As Boolean = True
        Private _flashTextDelay As Integer = 20
        Private _flashTextWrite As String = "Kernel Simulator"
        Private _flashTextBackgroundColor As String = New Color(ConsoleColor.Black).PlainSequence
        Private _flashTextMinimumRedColorLevel As Integer = 0
        Private _flashTextMinimumGreenColorLevel As Integer = 0
        Private _flashTextMinimumBlueColorLevel As Integer = 0
        Private _flashTextMinimumColorLevel As Integer = 0
        Private _flashTextMaximumRedColorLevel As Integer = 255
        Private _flashTextMaximumGreenColorLevel As Integer = 255
        Private _flashTextMaximumBlueColorLevel As Integer = 255
        Private _flashTextMaximumColorLevel As Integer = 0

        ''' <summary>
        ''' [FlashText] Enable 255 color support. Has a higher priority than 16 color support.
        ''' </summary>
        Public Property FlashText255Colors As Boolean
            Get
                Return _flashText255Colors
            End Get
            Set(value As Boolean)
                _flashText255Colors = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] Enable truecolor support. Has a higher priority than 255 color support.
        ''' </summary>
        Public Property FlashTextTrueColor As Boolean
            Get
                Return _flashTextTrueColor
            End Get
            Set(value As Boolean)
                _flashTextTrueColor = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] How many milliseconds to wait before making the next write?
        ''' </summary>
        Public Property FlashTextDelay As Integer
            Get
                Return _flashTextDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 20
                _flashTextDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] Text for FlashText. Shorter is better.
        ''' </summary>
        Public Property FlashTextWrite As String
            Get
                Return _flashTextWrite
            End Get
            Set(value As String)
                _flashTextWrite = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] Screensaver background color
        ''' </summary>
        Public Property FlashTextBackgroundColor As String
            Get
                Return _flashTextBackgroundColor
            End Get
            Set(value As String)
                _flashTextBackgroundColor = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] The minimum red color level (true color)
        ''' </summary>
        Public Property FlashTextMinimumRedColorLevel As Integer
            Get
                Return _flashTextMinimumRedColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _flashTextMinimumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] The minimum green color level (true color)
        ''' </summary>
        Public Property FlashTextMinimumGreenColorLevel As Integer
            Get
                Return _flashTextMinimumGreenColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _flashTextMinimumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] The minimum blue color level (true color)
        ''' </summary>
        Public Property FlashTextMinimumBlueColorLevel As Integer
            Get
                Return _flashTextMinimumBlueColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _flashTextMinimumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] The minimum color level (255 colors or 16 colors)
        ''' </summary>
        Public Property FlashTextMinimumColorLevel As Integer
            Get
                Return _flashTextMinimumColorLevel
            End Get
            Set(value As Integer)
                Dim FinalMinimumLevel As Integer = If(_flashText255Colors Or _flashTextTrueColor, 255, 15)
                If value <= 0 Then value = 0
                If value > FinalMinimumLevel Then value = FinalMinimumLevel
                _flashTextMinimumColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] The maximum red color level (true color)
        ''' </summary>
        Public Property FlashTextMaximumRedColorLevel As Integer
            Get
                Return _flashTextMaximumRedColorLevel
            End Get
            Set(value As Integer)
                If value <= _flashTextMinimumRedColorLevel Then value = _flashTextMinimumRedColorLevel
                If value > 255 Then value = 255
                _flashTextMaximumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] The maximum green color level (true color)
        ''' </summary>
        Public Property FlashTextMaximumGreenColorLevel As Integer
            Get
                Return _flashTextMaximumGreenColorLevel
            End Get
            Set(value As Integer)
                If value <= _flashTextMinimumGreenColorLevel Then value = _flashTextMinimumGreenColorLevel
                If value > 255 Then value = 255
                _flashTextMaximumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] The maximum blue color level (true color)
        ''' </summary>
        Public Property FlashTextMaximumBlueColorLevel As Integer
            Get
                Return _flashTextMaximumBlueColorLevel
            End Get
            Set(value As Integer)
                If value <= _flashTextMinimumBlueColorLevel Then value = _flashTextMinimumBlueColorLevel
                If value > 255 Then value = 255
                _flashTextMaximumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FlashText] The maximum color level (255 colors or 16 colors)
        ''' </summary>
        Public Property FlashTextMaximumColorLevel As Integer
            Get
                Return _flashTextMaximumColorLevel
            End Get
            Set(value As Integer)
                Dim FinalMaximumLevel As Integer = If(_flashText255Colors Or _flashTextTrueColor, 255, 15)
                If value <= _flashTextMinimumColorLevel Then value = _flashTextMinimumColorLevel
                If value > FinalMaximumLevel Then value = FinalMaximumLevel
                _flashTextMaximumColorLevel = value
            End Set
        End Property

    End Module

    Public Class FlashTextDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random
        Private CurrentWindowWidth As Integer
        Private CurrentWindowHeight As Integer
        Private ResizeSyncing As Boolean
        Private Left, Top As Integer

        Public Overrides Property ScreensaverName As String = "FlashText" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
            SetConsoleColor(New Color(FlashTextBackgroundColor), True)
            ConsoleWrapper.Clear()
            Wdbg(DebugLevel.I, "Console geometry: {0}x{1}", ConsoleWrapper.WindowWidth, ConsoleWrapper.WindowHeight)

            'Select position
            Left = RandomDriver.Next(ConsoleWrapper.WindowWidth)
            Top = RandomDriver.Next(ConsoleWrapper.WindowHeight)
            WdbgConditional(ScreensaverDebug, DebugLevel.I, "Selected left and top: {0}, {1}", Left, Top)
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            ConsoleWrapper.CursorVisible = False

            'Make two delay halves to make up one half for screen with text and one half for screen with no text to make a flashing effect
            Dim HalfDelay As Integer = FlashTextDelay / 2

            'Make a flashing text
            Console.BackgroundColor = ConsoleColor.Black
            ConsoleWrapper.Clear()
            If FlashTextTrueColor Then
                Dim RedColorNum As Integer = RandomDriver.Next(FlashTextMinimumRedColorLevel, FlashTextMaximumRedColorLevel)
                Dim GreenColorNum As Integer = RandomDriver.Next(FlashTextMinimumGreenColorLevel, FlashTextMaximumGreenColorLevel)
                Dim BlueColorNum As Integer = RandomDriver.Next(FlashTextMinimumBlueColorLevel, FlashTextMaximumBlueColorLevel)
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Got color (R;G;B: {0};{1};{2})", RedColorNum, GreenColorNum, BlueColorNum)
                Dim ColorStorage As New Color(RedColorNum, GreenColorNum, BlueColorNum)
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    WriteWhere(FlashTextWrite, Left, Top, True, ColorStorage)
                End If
            ElseIf FlashText255Colors Then
                Dim ColorNum As Integer = RandomDriver.Next(FlashTextMinimumColorLevel, FlashTextMaximumColorLevel)
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Got color ({0})", ColorNum)
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    WriteWhere(FlashTextWrite, Left, Top, True, New Color(ColorNum))
                End If
            Else
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    WriteWhere(FlashTextWrite, Left, Top, True, colors(RandomDriver.Next(FlashTextMinimumColorLevel, FlashTextMaximumColorLevel)))
                End If
            End If
            SleepNoBlock(HalfDelay, ScreensaverDisplayerThread)
            Console.BackgroundColor = ConsoleColor.Black
            ConsoleWrapper.Clear()
            SleepNoBlock(HalfDelay, ScreensaverDisplayerThread)

            'Reset resize sync
            ResizeSyncing = False
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
        End Sub

    End Class
End Namespace

File: Kernel Simulator/Misc/Screensaver/Displays/FlashText.cs
﻿//
// Kernel Simulator  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Kernel Simulator
//
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using Terminaux.Writer.ConsoleWriters;
using KS.Misc.Writers.DebugWriters;
using KS.Misc.Threading;
using Terminaux.Colors;
using Terminaux.Base;
using Terminaux.Colors.Data;
using KS.Misc.Reflection;

namespace KS.Misc.Screensaver.Displays
{
    /// <summary>
    /// Settings for FlashText
    /// </summary>
    public static class FlashTextSettings
    {
        private static bool _flashText255Colors;
        private static bool _flashTextTrueColor = true;
        private static int _flashTextDelay = 20;
        private static string _flashTextWrite = "Kernel Simulator";
        private static string _flashTextBackgroundColor = new Color(ConsoleColor.Black).PlainSequence;
        private static int _flashTextMinimumRedColorLevel = 0;
        private static int _flashTextMinimumGreenColorLevel = 0;
        private static int _flashTextMinimumBlueColorLevel = 0;
        private static int _flashTextMinimumColorLevel = 0;
        private static int _flashTextMaximumRedColorLevel = 255;
        private static int _flashTextMaximumGreenColorLevel = 255;
        private static int _flashTextMaximumBlueColorLevel = 255;
        private static int _flashTextMaximumColorLevel = 0;

        /// <summary>
        /// [FlashText] Enable 255 color support. Has a higher priority than 16 color support.
        /// </summary>
        public static bool FlashText255Colors
        {
            get
            {
                return _flashText255Colors;
            }
            set
            {
                _flashText255Colors = value;
            }
        }
        /// <summary>
        /// [FlashText] Enable truecolor support. Has a higher priority than 255 color support.
        /// </summary>
        public static bool FlashTextTrueColor
        {
            get
            {
                return _flashTextTrueColor;
            }
            set
            {
                _flashTextTrueColor = value;
            }
        }
        /// <summary>
        /// [FlashText] How many milliseconds to wait before making the next write?
        /// </summary>
        public static int FlashTextDelay
        {
            get
            {
                return _flashTextDelay;
            }
            set
            {
                if (value <= 0)
                    value = 20;
                _flashTextDelay = value;
            }
        }
        /// <summary>
        /// [FlashText] Text for FlashText. Shorter is better.
        /// </summary>
        public static string FlashTextWrite
        {
            get
            {
                return _flashTextWrite;
            }
            set
            {
                _flashTextWrite = value;
            }
        }
        /// <summary>
        /// [FlashText] Screensaver background color
        /// </summary>
        public static string FlashTextBackgroundColor
        {
            get
            {
                return _flashTextBackgroundColor;
            }
            set
            {
                _flashTextBackgroundColor = value;
            }
        }
        /// <summary>
        /// [FlashText] The minimum red color level (true color)
        /// </summary>
        public static int FlashTextMinimumRedColorLevel
        {
            get
            {
                return _flashTextMinimumRedColorLevel;
            }
            set
            {
                if (value <= 0)
                    value = 0;
                if (value > 255)
                    value = 255;
                _flashTextMinimumRedColorLevel = value;
            }
        }
        /// <summary>
        /// [FlashText] The minimum green color level (true color)
        /// </summary>
        public static int FlashTextMinimumGreenColorLevel
        {
            get
            {
                return _flashTextMinimumGreenColorLevel;
            }
            set
            {
                if (value <= 0)
                    value = 0;
                if (value > 255)
                    value = 255;
                _flashTextMinimumGreenColorLevel = value;
            }
        }
        /// <summary>
        /// [FlashText] The minimum blue color level (true color)
        /// </summary>
        public static int FlashTextMinimumBlueColorLevel
        {
            get
            {
                return _flashTextMinimumBlueColorLevel;
            }
            set
            {
                if (value <= 0)
                    value = 0;
                if (value > 255)
                    value = 255;
                _flashTextMinimumBlueColorLevel = value;
            }
        }
        /// <summary>
        /// [FlashText] The minimum color level (255 colors or 16 colors)
        /// </summary>
        public static int FlashTextMinimumColorLevel
        {
            get
            {
                return _flashTextMinimumColorLevel;
            }
            set
            {
                int FinalMinimumLevel = _flashText255Colors | _flashTextTrueColor ? 255 : 15;
                if (value <= 0)
                    value = 0;
                if (value > FinalMinimumLevel)
                    value = FinalMinimumLevel;
                _flashTextMinimumColorLevel = value;
            }
        }
        /// <summary>
        /// [FlashText] The maximum red color level (true color)
        /// </summary>
        public static int FlashTextMaximumRedColorLevel
        {
            get
            {
                return _flashTextMaximumRedColorLevel;
            }
            set
            {
                if (value <= _flashTextMinimumRedColorLevel)
                    value = _flashTextMinimumRedColorLevel;
                if (value > 255)
                    value = 255;
                _flashTextMaximumRedColorLevel = value;
            }
        }
        /// <summary>
        /// [FlashText] The maximum green color level (true color)
        /// </summary>
        public static int FlashTextMaximumGreenColorLevel
        {
            get
            {
                return _flashTextMaximumGreenColorLevel;
            }
            set
            {
                if (value <= _flashTextMinimumGreenColorLevel)
                    value = _flashTextMinimumGreenColorLevel;
                if (value > 255)
                    value = 255;
                _flashTextMaximumGreenColorLevel = value;
            }
        }
        /// <summary>
        /// [FlashText] The maximum blue color level (true color)
        /// </summary>
        public static int FlashTextMaximumBlueColorLevel
        {
            get
            {
                return _flashTextMaximumBlueColorLevel;
            }
            set
            {
                if (value <= _flashTextMinimumBlueColorLevel)
                    value = _flashTextMinimumBlueColorLevel;
                if (value > 255)
                    value = 255;
                _flashTextMaximumBlueColorLevel = value;
            }
        }
        /// <summary>
        /// [FlashText] The maximum color level (255 colors or 16 colors)
        /// </summary>
        public static int FlashTextMaximumColorLevel
        {
            get
            {
                return _flashTextMaximumColorLevel;
            }
            set
            {
                int FinalMaximumLevel = _flashText255Colors | _flashTextTrueColor ? 255 : 15;
                if (value <= _flashTextMinimumColorLevel)
                    value = _flashTextMinimumColorLevel;
                if (value > FinalMaximumLevel)
                    value = FinalMaximumLevel;
                _flashTextMaximumColorLevel = value;
            }
        }
    }

    /// <summary>
    /// Display code for FlashText
    /// </summary>
    public class FlashTextDisplay : BaseScreensaver, IScreensaver
    {

        private int Left, Top;

        /// <inheritdoc/>
        public override string ScreensaverName { get; set; } = "FlashText";

        /// <inheritdoc/>
        public override void ScreensaverPreparation()
        {
            // Variable preparations
            ColorTools.LoadBackDry(new Color(FlashTextSettings.FlashTextBackgroundColor));
            DebugWriter.Wdbg(DebugLevel.I, "Console geometry: {0}x{1}", ConsoleWrapper.WindowWidth, ConsoleWrapper.WindowHeight);

            // Select position
            Left = RandomDriver.RandomIdx(ConsoleWrapper.WindowWidth - FlashTextSettings.FlashTextWrite.Length);
            Top = RandomDriver.RandomIdx(ConsoleWrapper.WindowHeight);
            DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Selected left and top: {0}, {1}", Left, Top);
        }

        /// <inheritdoc/>
        public override void ScreensaverLogic()
        {
            ConsoleWrapper.CursorVisible = false;

            // Make two delay halves to make up one half for screen with text and one half for screen with no text to make a flashing effect
            int HalfDelay = (int)Math.Round(FlashTextSettings.FlashTextDelay / 2d);

            // Make a flashing text
            if (FlashTextSettings.FlashTextTrueColor)
            {
                int RedColorNum = RandomDriver.Random(FlashTextSettings.FlashTextMinimumRedColorLevel, FlashTextSettings.FlashTextMaximumRedColorLevel);
                int GreenColorNum = RandomDriver.Random(FlashTextSettings.FlashTextMinimumGreenColorLevel, FlashTextSettings.FlashTextMaximumGreenColorLevel);
                int BlueColorNum = RandomDriver.Random(FlashTextSettings.FlashTextMinimumBlueColorLevel, FlashTextSettings.FlashTextMaximumBlueColorLevel);
                DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Got color (R;G;B: {0};{1};{2})", RedColorNum, GreenColorNum, BlueColorNum);
                var ColorStorage = new Color(RedColorNum, GreenColorNum, BlueColorNum);
                if (!ConsoleResizeHandler.WasResized(false))
                {
                    TextWriterWhereColor.WriteWhereColorBack(FlashTextSettings.FlashTextWrite, Left, Top, true, ColorStorage, FlashTextSettings.FlashTextBackgroundColor);
                }
            }
            else
            {
                int ColorNum = RandomDriver.Random(FlashTextSettings.FlashTextMinimumColorLevel, FlashTextSettings.FlashTextMaximumColorLevel);
                DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Got color ({0})", ColorNum);
                if (!ConsoleResizeHandler.WasResized(false))
                {
                    TextWriterWhereColor.WriteWhereColorBack(FlashTextSettings.FlashTextWrite, Left, Top, true, new Color(ColorNum), FlashTextSettings.FlashTextBackgroundColor);
                }
            }
            ThreadManager.SleepNoBlock(HalfDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
            ColorTools.LoadBackDry(new Color(ConsoleColors.Black));
            ThreadManager.SleepNoBlock(HalfDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);

            // Reset resize sync
            ConsoleResizeHandler.WasResized();
        }

    }
}